_posts/2019-01-24-how-to-keep-safe-as-a-client.md

```diff
@@ -0,0 +1,129 @@
+//Invoke remote service here
+
+```
+### Try to reconnect
+
+After cut off connections, we need to recover as soon as possible. Because currently the system is not complete. In order to automatically recover connection we need to define a standard, like previously how we judge a system is unavailable, we need to define whether a system is available again.
+- Threshold: for example, there are 100 success requests within 5 seconds.
+- Percentage: for example, there are 95% success requests within 10 seconds.
+
+It also contains time window, threshold and percentage. But there is a slight difference: most of the time, the system will keep unavailable for a moment, So we can check in a fixed times(30s for example).
+
+```python
+successCount = 0;
+isHalfOpen = true;
+
+if(success){
+  if(isHalfOpen){
+    successCount ++;
+    if(successCount == threshold){
+      isOpenCircuitBreaker = false;
+    }
+  }
+  return success;
+}else{
+  errorCount ++;
+  if(errorCount == threshold){
+    isOpenCircuitBreaker = true;
+  }
+}
+
+```
+However reconnect may failed, in a large system we can't use all servers to try reconnecting. We can take these measures:
+- 1.Use a few servers to check
+- 2.Add a health check interface to check system conditions: CPU, IO etc.
+
+### recover
+
+Once the system is available, we need to recover.
+
+![](https://mmbiz.qpic.cn/mmbiz_png/oB5bd6W6hI31YcnTS8xgHlND7GJnk9vSUGPwicRFU6xJeNPxzKOWEhlM3yKSVdEfk1nusm5FfnRm1WyYkOVybBg/640?wx_fmt=png&tp=webp&wxfrom=5&wx_lazy=1&wx_co=1)
+
+This is the general process of Circuit Breaker, frameworks like hystrix, resilience4j are all have such process flow.
+
+## Best practice for Circuit Breaker
+
+### What kind of situation require Circuit Breaker
+
+Circuit Breaker is not suitable in every situations. There are some conditions we can consider using Circuit Breaker:
+- The system we rely on is a shared system, and our client is just one of them. Other systems may abuse the shared system and affect us.
+- Our system is not used in a regular base, somethings the system may used very frequently while other times it is rarely used, we need to hold on the sudden flow.
+
+### Pay attention to implement a Circuit Breaker
+
+Circuit Breaker is not perfect, we need to pay attention to two things.
+- Firstly, if we rely on a system that has many replicas, we need to keep in mind that exception in partial node does not mean exception of the whole system.
+- Secondly, Circuit Breaker should be the last choice, we can to use "fall back" or "rate limit" prior than Circuit Breaker. Because partial is better than none, although some users may find out service broke, we can try to keep our system working and serve other users. For example, abandon non-core services, or give a friendly hint.
+
+## Conclusion
+
+Circuit Breaker can be implemented before actual function or after actual function. It is very correspond to AOP concept, so generally most circuit breaker framework use AOP as an implement strategy.
```

Comment:
Please give the full phrase when you use acronyms for the first time to avoid confusion.
For example, the first time I use OOP (object-oriented programming) in my article, I'd put the full phrase in parenthesis.
circuit breaker framework => circuit breaker ```frameworks```